File: internal/models/subdomain.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// SubdomainEnumTask 子域名枚举任务
type SubdomainEnumTask struct {
	ID            primitive.ObjectID   `bson:"_id,omitempty" json:"id"`
	ProjectID     primitive.ObjectID   `bson:"projectId" json:"projectId"`
	RootDomain    string               `bson:"rootDomain" json:"rootDomain"`
	TaskName      string               `bson:"taskName" json:"taskName"`
	Status        string               `bson:"status" json:"status"` // pending, running, completed, failed
	CreatedAt     time.Time            `bson:"createdAt" json:"createdAt"`
	StartedAt     time.Time            `bson:"startedAt" json:"startedAt"`
	CompletedAt   time.Time            `bson:"completedAt" json:"completedAt"`
	Progress      float64              `bson:"progress" json:"progress"` // 0-100
	Config        SubdomainEnumConfig  `bson:"config" json:"config"`
	ResultSummary SubdomainEnumSummary `bson:"resultSummary" json:"resultSummary"`
	Error         string               `bson:"error" json:"error"`
	NodeID        string               `bson:"nodeId" json:"nodeId"` // 执行任务的节点ID
	Tags          []string             `bson:"tags" json:"tags"`
}

// SubdomainEnumConfig 子域名枚举配置
type SubdomainEnumConfig struct {
	Methods           []string `bson:"methods" json:"methods"`                     // 枚举方法: dns_brute, dns_zone_transfer, search_engines, certificate_transparency, etc.
	DictionaryPath    string   `bson:"dictionaryPath" json:"dictionaryPath"`       // 字典路径
	Concurrency       int      `bson:"concurrency" json:"concurrency"`             // 并发数
	Timeout           int      `bson:"timeout" json:"timeout"`                     // 超时时间(秒)
	RetryCount        int      `bson:"retryCount" json:"retryCount"`               // 重试次数
	ResolverServers   []string `bson:"resolverServers" json:"resolverServers"`     // DNS解析服务器
	RateLimit         int      `bson:"rateLimit" json:"rateLimit"`                 // 请求速率限制(每秒)
	IncludeWildcard   bool     `bson:"includeWildcard" json:"includeWildcard"`     // 是否包含泛解析域名
	VerifySubdomains  bool     `bson:"verifySubdomains" json:"verifySubdomains"`   // 是否验证子域名
	SaveToDB          bool     `bson:"saveToDB" json:"saveToDB"`                   // 是否保存到数据库
	RecursiveSearch   bool     `bson:"recursiveSearch" json:"recursiveSearch"`     // 是否递归搜索
	RecursiveDepth    int      `bson:"recursiveDepth" json:"recursiveDepth"`       // 递归深度
	IncludeThirdLevel bool     `bson:"includeThirdLevel" json:"includeThirdLevel"` // 是否包含三级域名
}

// SubdomainEnumSummary 子域名枚举结果摘要
type SubdomainEnumSummary struct {
	TotalFound       int            `bson:"totalFound" json:"totalFound"`             // 发现的子域名总数
	NewFound         int            `bson:"newFound" json:"newFound"`                 // 新发现的子域名数
	ResolvedCount    int            `bson:"resolvedCount" json:"resolvedCount"`       // 可解析的子域名数
	UnresolvedCount  int            `bson:"unresolvedCount" json:"unresolvedCount"`   // 不可解析的子域名数
	WildcardCount    int            `bson:"wildcardCount" json:"wildcardCount"`       // 泛解析域名数
	TakeOverCount    int            `bson:"takeOverCount" json:"takeOverCount"`       // 可能被接管的子域名数
	MethodStats      map[string]int `bson:"methodStats" json:"methodStats"`           // 各方法发现的子域名数
	ProcessedRecords int            `bson:"processedRecords" json:"processedRecords"` // 处理的记录数
}

// SubdomainResult 子域名枚举结果
type SubdomainResult struct {
	ID           primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	TaskID       primitive.ObjectID `bson:"taskId" json:"taskId"`
	ProjectID    primitive.ObjectID `bson:"projectId" json:"projectId"`
	RootDomain   string             `bson:"rootDomain" json:"rootDomain"`
	Subdomain    string             `bson:"subdomain" json:"subdomain"`
	IPs          []string           `bson:"ips" json:"ips"`
	CNAME        string             `bson:"cname" json:"cname"`
	Type         string             `bson:"type" json:"type"` // A, AAAA, CNAME, etc.
	Records      []DNSRecord        `bson:"records" json:"records"`
	IsWildcard   bool               `bson:"isWildcard" json:"isWildcard"`
	IsResolved   bool               `bson:"isResolved" json:"isResolved"`
	IsTakeOver   bool               `bson:"isTakeOver" json:"isTakeOver"`
	TakeOverType string             `bson:"takeOverType" json:"takeOverType"`
	Source       string             `bson:"source" json:"source"` // 发现方法
	CreatedAt    time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt    time.Time          `bson:"updatedAt" json:"updatedAt"`
	AssetID      primitive.ObjectID `bson:"assetId" json:"assetId"` // 关联的资产ID
}

// DNSRecord DNS记录
type DNSRecord struct {
	Type  string `bson:"type" json:"type"`   // A, AAAA, CNAME, TXT, MX, NS, etc.
	Value string `bson:"value" json:"value"` // 记录值
	TTL   int    `bson:"ttl" json:"ttl"`     // 生存时间
}

// SubdomainDictionary 子域名字典
type SubdomainDictionary struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Name        string             `bson:"name" json:"name"`
	Description string             `bson:"description" json:"description"`
	Count       int                `bson:"count" json:"count"`
	CreatedAt   time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt   time.Time          `bson:"updatedAt" json:"updatedAt"`
	FilePath    string             `bson:"filePath" json:"filePath"`
	IsDefault   bool               `bson:"isDefault" json:"isDefault"`
	Tags        []string           `bson:"tags" json:"tags"`
}

// SubdomainTakeOverRule 子域名接管规则
type SubdomainTakeOverRule struct {
	ID           primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Service      string             `bson:"service" json:"service"`           // 服务名称
	Fingerprint  string             `bson:"fingerprint" json:"fingerprint"`   // 指纹
	CNAMEPattern string             `bson:"cnamePattern" json:"cnamePattern"` // CNAME模式
	Status       string             `bson:"status" json:"status"`             // 状态: active, inactive
	Description  string             `bson:"description" json:"description"`   // 描述
	Reference    string             `bson:"reference" json:"reference"`       // 参考链接
	CreatedAt    time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt    time.Time          `bson:"updatedAt" json:"updatedAt"`
}

File: web/src/lib/types/subdomain.ts
/**
 * 子域名枚举相关类型定义
 */

/**
 * 子域名枚举任务配置
 */
export interface SubdomainConfig {
	target?: string; // 目标域名
	dictionaryPath?: string; // 字典文件路径
	wordlistPath?: string; // 字典文件路径（别名）
	methods?: string[]; // 枚举方法列表
	enumMethods?: string[]; // 枚举方法列表（别名）
	concurrency?: number; // 最大并发数
	maxWorkers?: number; // 最大并发数（别名）
	timeout?: number; // 超时时间（秒）
	retryCount?: number; // 最大重试次数
	rateLimit?: number; // 速率限制（每秒请求数）
	resolverServers?: string[]; // DNS服务器列表
	dnsServers?: string[]; // DNS服务器列表（别名）
	verifySubdomains?: boolean; // 验证子域名活跃性
	recursiveSearch?: boolean; // 启用递归枚举
	enableWildcard?: boolean; // 启用通配符检测
	saveToDB?: boolean; // 保存到数据库
}

/**
 * 子域名枚举任务
 */
export interface SubdomainTask {
	id: string;
	name: string;
	description?: string;
	type: 'subdomain_enum';
	status: 'pending' | 'queued' | 'running' | 'completed' | 'failed' | 'canceled';
	priority: number;
	config: SubdomainConfig;
	rootDomain: string; // 根域名
	projectId?: string;
	nodeId?: string;
	createdBy: string;
	createdAt: string;
	updatedAt: string;
	startTime?: string;
	endTime?: string;
	progress?: TaskProgress;
	result?: TaskResult;
	error?: string;
	tags?: string[];
}

/**
 * 任务进度
 */
export interface TaskProgress {
	total: number;
	current: number;
	completed: number; // 已完成数量
	found: number;
	failed: number;
	success: number; // 成功数量
	percentage: number;
	speed?: number; // 每秒处理数量
	estimatedTime?: number; // 预计剩余时间（秒）
}

/**
 * 任务结果
 */
export interface TaskResult {
	subdomains: SubdomainResult[];
	uniqueSubdomains: SubdomainResult[];
	subdomainCount: number;
	uniqueCount: number;
	target: string;
	enumMethods: string[];
	sourceStats: Record<string, number>;
	scanDuration: number;
	summary: string;
}

/**
 * 子域名枚举结果
 */
export interface SubdomainResult {
	subdomain: string;
	domain?: string; // 添加域名字段
	ips: string[];
	ip?: string; // 主要IP地址
	cname?: string;
	status: 'found' | 'not_found';
	source: string;
	responseTime: number;
	httpStatus?: number;
	httpTitle?: string;
	technologies?: string[];
	takeover?: TakeoverInfo;
	timestamp: string;
	createdAt?: string; // 添加创建时间字段
	metadata?: Record<string, string>;
}

/**
 * 子域名接管信息
 */
export interface TakeoverInfo {
	vulnerable: boolean;
	service?: string;
	pattern?: string;
	cname?: string;
}

/**
 * 任务创建请求
 */
export interface SubdomainTaskCreateRequest {
	name: string;
	target: string;
	projectId?: string;
	wordlistPath?: string;
	enumMethods?: string[];
	maxWorkers?: number;
	timeout?: number;
	maxRetries?: number;
	rateLimit?: number;
	dnsServers?: string[];
	verifySubdomains?: boolean;
	enableRecursive?: boolean;
	tags?: string[];
	advanced?: {
		maxWorkers?: number;
		timeout?: number;
		maxRetries?: number;
		rateLimit?: number;
		enableWildcard?: boolean;
		enableDoH?: boolean;
		enableRecursive?: boolean;
		maxDepth?: number;
		verifySubdomains?: boolean;
		enableCache?: boolean;
		cacheTimeout?: number;
	};
}

/**
 * 任务列表响应
 */
export interface TaskListResponse {
	tasks: SubdomainTask[];
	total: number;
	page: number;
	limit: number;
	totalPages: number;
}

/**
 * 子域名枚举表单数据
 */
export interface SubdomainFormData {
	name: string;
	description: string;
	target: string;
	methodPreset: string;
	enumMethods: string[];
	wordlistType: 'preset' | 'custom';
	wordlistPreset: string;
	customWordlist: string[];
	dnsPreset: string;
	customDnsServers: string[];
	advanced: {
		maxWorkers: number;
		timeout: number;
		maxRetries: number;
		rateLimit: number;
		enableWildcard: boolean;
		enableDoH: boolean;
		enableRecursive: boolean;
		maxDepth: number;
		verifySubdomains: boolean;
		enableCache: boolean;
		cacheTimeout: number;
	};
	searchEngines: {
		enableGoogle: boolean;
		enableBing: boolean;
		censysApiKey: string;
		shodanApiKey: string;
	};
	projectId?: string;
}

/**
 * 子域名枚举统计数据
 */
export interface SubdomainStatistics {
	totalTasks: number;
	completedTasks: number;
	failedTasks: number;
	totalSubdomains: number;
	uniqueSubdomains: number;
	successRate: number;
	averageDuration: number;
	sourceStats: Record<string, number>;
	methodStats: Record<string, number>;
	recentTasks: SubdomainTask[];
	topTargets: Array<{
		target: string;
		count: number;
		lastScan: string;
	}>;
}

/**
 * 枚举方法预设
 */
export interface EnumMethodPreset {
	id: string;
	name: string;
	methods: string[];
	description: string;
	estimatedTime: string;
}

/**
 * DNS服务器预设
 */
export interface DNSPreset {
	id: string;
	name: string;
	servers: string[];
	description: string;
	location?: string;
}

/**
 * 字典预设
 */
export interface WordlistPreset {
	id: string;
	name: string;
	size: string;
	description: string;
	category: 'common' | 'comprehensive' | 'security' | 'custom';
}

/**
 * 扫描任务过滤器
 */
export interface SubdomainFilter {
	status?: string;
	target?: string;
	projectId?: string;
	dateRange?: {
		startDate: string;
		endDate: string;
	};
	createdBy?: string;
	enumMethod?: string;
}

/**
 * 子域名枚举配置验证结果
 */
export interface ValidationResult {
	valid: boolean;
	errors: string[];
	warnings: string[];
	estimatedDuration?: number;
	wordlistSize?: number;
}

/**
 * 实时进度更新事件
 */
export interface ProgressUpdateEvent {
	taskId: string;
	progress: TaskProgress;
	latestResults?: SubdomainResult[];
	status: string;
}

/**
 * 导出配置
 */
export interface ExportConfig {
	format: 'csv' | 'json' | 'xlsx' | 'xml';
	includeDetails: boolean;
	includeMetadata: boolean;
	onlyActive: boolean;
	groupBySource: boolean;
	customFields?: string[];
}

/**
 * 子域名详细信息
 */
export interface SubdomainDetail {
	subdomain: string;
	ips: string[];
	cname?: string;
	status: string;
	source: string;
	httpInfo?: {
		status: number;
		title?: string;
		server?: string;
		technologies: string[];
		responseTime: number;
	};
	sslInfo?: {
		issuer: string;
		subject: string;
		notBefore: string;
		notAfter: string;
		fingerprint: string;
	};
	dnsInfo?: {
		aRecords: string[];
		cnameRecord?: string;
		mxRecords: string[];
		txtRecords: string[];
		nsRecords: string[];
	};
	geoInfo?: {
		country: string;
		region: string;
		city: string;
		org: string;
		asn: string;
	};
	takeoverInfo?: TakeoverInfo;
	firstSeen: string;
	lastSeen: string;
	scanCount: number;
}

/**
 * 子域名发现来源统计
 */
export interface SourceStatistics {
	source: string;
	count: number;
	percentage: number;
	averageResponseTime: number;
	successRate: number;
}

/**
 * 子域名验证结果
 */
export interface SubdomainVerification {
	subdomain: string;
	isActive: boolean;
	httpStatus?: number;
	httpsStatus?: number;
	responseTime: number;
	technologies: string[];
	lastChecked: string;
}

/**
 * 递归枚举配置
 */
export interface RecursiveConfig {
	enabled: boolean;
	maxDepth: number;
	prefixes: string[];
	skipWildcard: boolean;
	onlyActive: boolean;
}

/**
 * 搜索引擎配置
 */
export interface SearchEngineConfig {
	google: {
		enabled: boolean;
		apiKey?: string;
		searchEngineId?: string;
		maxResults: number;
	};
	bing: {
		enabled: boolean;
		apiKey?: string;
		maxResults: number;
	};
	censys: {
		enabled: boolean;
		apiId?: string;
		apiSecret?: string;
	};
	shodan: {
		enabled: boolean;
		apiKey?: string;
	};
}
